File: older-code/inaimathi/secret-santa--aka-dann-assigns-an-NP-complete-problem/santa.pl
%% people(['Alex', 'Beatrice', 'Clarence', 'Diego', 'Ernest', 'Francesca', 'Gwen', 'Heidi', 'Ingrid']).

person('Alex').
person('Beatrice').
person('Clarence').
person('Diego').
person('Ernest').
person('Francesca').
person('Gwen').
person('Heidi').
person('Ingrid').

giving(A, B) :- 
    person(A),
    person(B),
    A \= B.

something([A, B], [[A, B]]) :-
    giving(A, B).

something([A, B | Rest], [[A, B] | Pairings]) :-
    giving(A, B),
    something([B | Rest], Pairings),
    not(member([A, B], Pairings)),
    not(member([B, A], Pairings)).

File: older-code/inaimathi/secret-santa--aka-dann-assigns-an-NP-complete-problem/Santa.hs
module Santa where

import Data.List

names = ["Alex", "Beatrice", "Clarence", "Diego", "Ernest"] -- , "Francesca", "Gwen", "Heidi", "Ingrid"

-- lets_be_santa :: [String] -> [[(String, String)]]
lets_be_santa names = sequence . map (pair_with names) $ reverse names

pair_with :: [a] -> a -> [(a, a)]
pair_with lst elem = zip lst $ repeat elem

lets_be_picky_santa filter_fn names = take 1 . dropWhile filter_fn $ lets_be_santa names

is_selfie [] = False
is_selfie ((a, b):rest) 
    | a == b = True
    | otherwise = is_selfie rest 

is_multigiving pairs = (length pairs) /= (length . nub $ map fst pairs)

picky pairs = or [(is_selfie pairs), (is_multigiving pairs)]
